optimum/onnxruntime/modeling_decoder.py

```diff
@@ -0,0 +1,742 @@
+        provider_options: Optional[Dict] = None,
+        **kwargs
+    ):
+        """
+        Creates an instance of [`~optimum.onnxruntime.modeling_causal.ORTModelForCausalLM`].
+        Three inference sessions will be created for respectively the decoder and decoder with past key values
+        models. The default provider is `CPUExecutionProvider` to match the default behaviour in PyTorch/TensorFlow/JAX.
+
+        Arguments:
+            decoder_path (`str` or `Path`):
+                The path of the decoder ONNX model.
+            decoder_with_past_path (`str` or `Path`, *optional*):
+                The path of the decoder with past key values ONNX model.
+            provider(`str`, *optional*):
+                The ONNX Runtime provider to use for loading the model. Defaults to `"CPUExecutionProvider"`.
+            session_options (`onnxruntime.SessionOptions`, *optional*),:
+                ONNX Runtime session options to use for loading the model. Defaults to `None`.
+            provider_options (`Dict`, **optional**):
+                Provider option dictionary corresponding to the provider used. See available options
+                for each provider: https://onnxruntime.ai/docs/api/c/group___global.html . Defaults to `None`.
+        """
+        available_providers = onnxruntime.get_available_providers()
+        if provider not in available_providers:
+            raise ValueError(
+                f"Asked to use {provider} as an ONNX Runtime execution provider, but the available execution providers are {available_providers}."
+            )
+
+        providers = [provider]
+        if provider == "TensorrtExecutionProvider":
+            # follow advice in https://onnxruntime.ai/docs/execution-providers/TensorRT-ExecutionProvider.html#python
+            providers.append("CUDAExecutionProvider")
+
+        decoder_session = onnxruntime.InferenceSession(
+            str(decoder_path),
+            providers=providers,
+            sess_options=session_options,
+            provider_options=None if provider_options is None else [provider_options],
+        )
+        decoder_with_past_session = None
+        # If a decoder_with_past_path is provided, an inference session for the decoder with past key/values as inputs
+        # will be enabled
+        if decoder_with_past_path is not None:
+            decoder_with_past_session = onnxruntime.InferenceSession(
+                str(decoder_with_past_path),
+                providers=providers,
+                sess_options=session_options,
+                provider_options=None if provider_options is None else [provider_options],
+            )
+        return decoder_session, decoder_with_past_session
+
+    def _save_pretrained(
+        self,
+        save_directory: Union[str, Path],
+        decoder_file_name: Optional[str] = None,
+        decoder_with_past_file_name: Optional[str] = None,
+        **kwargs
```

Comment:
Should not have this.

Reply:
I don't disagree with this proposition but the modification should first be done in `OptimizedModel`, it makes sense to do this modification (and adapt the classes inheriting from it) in an other PR (like #490)